modeshape-jcr/src/main/java/org/modeshape/jcr/locking/StandaloneLockingService.java

```diff
@@ -0,0 +1,222 @@
+            boolean success = false;
+            try {
+                success = time > 0 ? lock.tryLock(time, unit) : lock.tryLock();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Thread " + Thread.currentThread().getName()
+                             + " received interrupt request while waiting to acquire lock '{0}'", name);
+                Thread.currentThread().interrupt();
+            }
+            if (!success) {
+                LOGGER.debug("Unable to acquire lock on {0}. Reverting back the already obtained locks: {1}", name,
+                             successfullyLocked);
+                synchronized (locksByName) {
+                    successfullyLocked.stream().forEach(lockName -> locksByName.get(lockName).unlock());
+                }
+                return false;
+            }
+
+            // we now have a lock but we need to make sure it hasn't just been marked for cleanup
+            // this may've been unlocked by another thread, so refresh to make sure a strong reference is active
+            validateLock(name, lock);
+
+            LOGGER.debug("{0} locked successfully", lock);
+            successfullyLocked.add(name);
+        }
+        return true;
+    }
+
+    private void validateLock(String name, NamedLock lock) {
+        if (!lock.hasStrongReference()) {
+            synchronized (locksByName) {
+                // we have to do a 'put' here because the key (name) may be a different object instance than the original one
+                // even if this name is 'equal' to the old name
+                // we don't change the lock instance, but we must wrap a new key instance
+                locksByName.put(name, lock.resetStrongReference(name));
+            }
+        }
+    }
+
+    @Override
+    public List<String> unlock(String... names) {
+        return Arrays.stream(names)
+                     .map(this::unlock)
+                     .filter(Optional::isPresent)
+                     .map(Optional::get)
+                     .collect(Collectors.toList());
+    }
+    
+    private Optional<String> unlock(String name) {
+        LOGGER.debug("Attempting to unlock {0}", name);
+        NamedLock lock = locksByName.get(name);
+        if (lock == null) {
+            return Optional.of(name);
+        }
+        try {
+            // this will null-out the strong reference to the key if successful, meaning the lock can be removed from 
+            // map at some point
+            lock.unlock();
+        } catch (IllegalMonitorStateException e) {
+            LOGGER.warn(JcrI18n.warnAnotherThreadAttemptingToUnlock, Thread.currentThread().getName(), lock);
+        }
+        if (lock.isLocked()) {
+            LOGGER.debug("Failed to unlock {0} ", lock);
+            return Optional.of(name);
+        } else {
+            LOGGER.debug("{0} unlocked", lock);
+            return Optional.empty();
+        }
+    }
+
+    private class NamedLock implements Lock {
```

Comment:
I'd mark this as `@ThreadSafe` since it is potentially used concurrently from multiple threads to lock and/or unlock it, and so the `name` field is kept to be modified atomically.